velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.cpp

```diff
@@ -26,30 +29,59 @@
 namespace facebook::velox::filesystems {
 
 #ifdef VELOX_ENABLE_S3
-folly::once_flag S3FSInstantiationFlag;
-
-// Only one instance of S3FileSystem is supported for now.
-// TODO: Support multiple S3FileSystem instances using a cache
-static std::shared_ptr<S3FileSystem> s3fs = nullptr;
+std::mutex mtx1, mtx2;
+/// Multiple S3 filesystems are supported.
+/// Key is the endpoint value specified in the config using hive.s3.endpoint.
+/// If the endpoint is empty, it will default to AWS S3.
+/// We cannot use folly::ConcurrentHashMap to store the filesystem instances.
+/// This is because we need to delete the filesystems during
+/// finalizeS3FileSystem and folly::ConcurrentHashMap does not guarantee the
+/// entries are cleared on invoking clear().
+static std::unordered_map<std::string, std::shared_ptr<FileSystem>> filesystems;
 
 std::function<std::shared_ptr<
     FileSystem>(std::shared_ptr<const Config>, std::string_view)>
 fileSystemGenerator() {
   static auto filesystemGenerator = [](std::shared_ptr<const Config> properties,
                                        std::string_view filePath) {
-    folly::call_once(S3FSInstantiationFlag, [&properties]() {
-      std::shared_ptr<S3FileSystem> fs;
-      if (properties != nullptr) {
-        initializeS3(properties.get());
-        fs = std::make_shared<S3FileSystem>(properties);
-      } else {
-        auto config = std::make_shared<core::MemConfig>();
-        initializeS3(config.get());
-        fs = std::make_shared<S3FileSystem>(config);
+    static folly::
+        ConcurrentHashMap<std::string, std::shared_ptr<folly::once_flag>>
+            s3InitiationFlags;
+    std::string s3Identity("aws-s3");
+    if (properties) {
+      auto endpoint = connector::hive::HiveConfig::s3Endpoint(properties.get());
+      if (!endpoint.empty()) {
+        // The identity is the endpoint.
+        s3Identity = endpoint;
+      }
+    }
+    std::unique_lock<std::mutex> lk(mtx1, std::defer_lock);
+    /// If the init flag for a given s3 endpoint is not found,
+    /// create one for init use. It's a singleton.
+    if (s3InitiationFlags.find(s3Identity) == s3InitiationFlags.end()) {
+      lk.lock();
```

Comment:
can we just instantiate a lock_guard inside this branch instead?

Reply:
fixed